File: nodecore-cli/src/main/java/nodecore/cli/serialization/BlockInfo.java
// VeriBlock NodeCore CLI
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.cli.serialization;

import com.google.gson.annotations.SerializedName;
import com.google.protobuf.ByteString;
import nodecore.api.grpc.VeriBlockMessages;
import nodecore.api.grpc.utilities.ByteStringUtility;
import org.veriblock.core.utilities.BlockUtility;
import org.veriblock.core.utilities.Utility;

import java.util.ArrayList;
import java.util.List;

public class BlockInfo {
    public BlockInfo(final VeriBlockMessages.Block block) {
        hash = ByteStringUtility.byteStringToHex(block.getHash());
        number = block.getNumber();
        version = (short)block.getVersion();
        previousHash = ByteStringUtility.byteStringToHex(block.getPreviousHash());
        if (number % 20 == 0) {
            previousKeystoneHash = ByteStringUtility.byteStringToHex(block.getSecondPreviousHash());
            secondPreviousKeystoneHash = ByteStringUtility.byteStringToHex(block.getThirdPreviousHash());
        } else {
            secondPreviousHash = ByteStringUtility.byteStringToHex(block.getSecondPreviousHash());
            thirdPreviousHash = ByteStringUtility.byteStringToHex(block.getThirdPreviousHash());
        }
        merkleRoot = ByteStringUtility.byteStringToHex(block.getMerkleRoot());
        timestamp = block.getTimestamp();
        encoded_difficulty = block.getEncodedDifficulty();
        winningNonce = block.getWinningNonce();

        for (VeriBlockMessages.TransactionUnion union : block.getRegularTransactionsList())
            if (union.getTransactionCase() == VeriBlockMessages.TransactionUnion.TransactionCase.SIGNED) {
                regular_transactions.add(ByteStringUtility.byteStringToHex(union.getSigned().getTransaction().getTxId()));
            }
            else {
                regular_transactions.add(ByteStringUtility.byteStringToHex(union.getSignedMultisig().getTransaction().getTxId()));
            }
        for (VeriBlockMessages.TransactionUnion union : block.getPopTransactionsList())
            pop_transactions.add(ByteStringUtility.byteStringToHex(union.getSigned().getTransaction().getTxId()));

        totalFees = Utility.formatAtomicLongWithDecimal(block.getTotalFees());
        powCoinbaseReward = Utility.formatAtomicLongWithDecimal(block.getPowCoinbaseReward());
        popCoinbaseReward = Utility.formatAtomicLongWithDecimal(block.getPopCoinbaseReward());

        for (ByteString header : block.getBitcoinBlockHeadersList())
            bitcoinBlockHeaders.add(ByteStringUtility.byteStringToHex(header));

        blockContentMetapackage = new BlockContentMetapackageInfo(block.getBlockContentMetapackage());
        size = block.getSize();

        if (number % 20 == 0) {
            header = Utility.bytesToHex(BlockUtility.assembleBlockHeader(number,
                    version,
                    previousHash,
                    previousKeystoneHash,
                    secondPreviousKeystoneHash,
                    merkleRoot,
                    timestamp,
                    encoded_difficulty,
                    winningNonce));
        } else {
            header = Utility.bytesToHex(BlockUtility.assembleBlockHeader(number,
                    version,
                    previousHash,
                    secondPreviousHash,
                    thirdPreviousHash,
                    merkleRoot,
                    timestamp,
                    encoded_difficulty,
                    winningNonce));
        }
    }

    @SerializedName("hash")
    public String hash;

    @SerializedName("header")
    public String header;

    @SerializedName("number")
    public int number;

    @SerializedName("version")
    public short version;

    @SerializedName("previous_hash")
    public String previousHash;

    @SerializedName("second_previous_hash")
    public String secondPreviousHash;

    @SerializedName("previous_keystone_hash")
    public String previousKeystoneHash;

    @SerializedName("third_previous_hash")
    public String thirdPreviousHash;

    @SerializedName("second_previous_keystone_hash")
    public String secondPreviousKeystoneHash;

    @SerializedName("merkle_root")
    public String merkleRoot;

    @SerializedName("timestamp")
    public int timestamp;

    @SerializedName("encoded_difficulty")
    public int encoded_difficulty;

    @SerializedName("winning_nonce")
    public int winningNonce;

    @SerializedName("bitcoin_block_headers")
    public List<String> bitcoinBlockHeaders = new ArrayList<>();

    @SerializedName("regular_transactions")
    public List<String> regular_transactions = new ArrayList<>();

    @SerializedName("pop_transactions")
    public List<String> pop_transactions = new ArrayList<>();

    @SerializedName("total_fees")
    public String totalFees;

    @SerializedName("pow_coinbase_reward")
    public String powCoinbaseReward;

    @SerializedName("pop_coinbase_reward")
    public String popCoinbaseReward;

    @SerializedName("block_content_metapackage")
    public BlockContentMetapackageInfo blockContentMetapackage;

    @SerializedName("size")
    public int size;
}

File: nodecore-cli/src/main/java/nodecore/cli/serialization/BlockInfo.kt
// VeriBlock NodeCore CLI
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.cli.serialization

import com.google.gson.annotations.SerializedName
import nodecore.api.grpc.VeriBlockMessages
import nodecore.api.grpc.VeriBlockMessages.TransactionUnion
import nodecore.api.grpc.utilities.ByteStringUtility
import nodecore.api.grpc.utilities.extensions.toHex
import org.veriblock.core.utilities.BlockUtility
import org.veriblock.core.utilities.Utility
import org.veriblock.core.utilities.extensions.formatAtomicLongWithDecimal
import org.veriblock.core.utilities.extensions.toHex

class BlockInfo(
    block: VeriBlockMessages.Block
) {
    @SerializedName("hash")
    val hash = block.hash.toHex()

    @SerializedName("number")
    val number = block.number

    @SerializedName("version")
    val version = block.version.toShort()

    @SerializedName("previous_hash")
    val previousHash = block.previousHash.toHex()

    @SerializedName("second_previous_hash")
    val secondPreviousHash = if (number % 20 == 0) {
        null
    } else {
        block.secondPreviousHash.toHex()
    }

    @SerializedName("previous_keystone_hash")
    val previousKeystoneHash = if (number % 20 == 0) {
        block.secondPreviousHash.toHex()
    } else {
        null
    }

    @SerializedName("third_previous_hash")
    val thirdPreviousHash = if (number % 20 == 0) {
        block.thirdPreviousHash.toHex()
    } else {
        null
    }

    @SerializedName("second_previous_keystone_hash")
    val secondPreviousKeystoneHash = if (number % 20 == 0) {
        block.thirdPreviousHash.toHex()
    } else {
        null
    }

    @SerializedName("merkle_root")
    val merkleRoot = block.merkleRoot.toHex()

    @SerializedName("timestamp")
    val timestamp = block.timestamp

    @SerializedName("encoded_difficulty")
    val encoded_difficulty = block.encodedDifficulty

    @SerializedName("winning_nonce")
    val winningNonce = block.winningNonce

    @SerializedName("bitcoin_block_headers")
    val bitcoinBlockHeaders = block.bitcoinBlockHeadersList.map {
        it.toHex()
    }

    @SerializedName("regular_transactions")
    val regular_transactions = block.regularTransactionsList.map { transactionUnion ->
        if (transactionUnion.transactionCase == TransactionUnion.TransactionCase.SIGNED) {
            transactionUnion.signed.transaction.txId.toHex()
        } else {
            transactionUnion.signedMultisig.transaction.txId.toHex()
        }
    }

    @SerializedName("pop_transactions")
    val pop_transactions = block.popTransactionsList.map { transactionUnion ->
        transactionUnion.signed.transaction.txId.toHex()
    }

    @SerializedName("total_fees")
    val totalFees = block.totalFees.formatAtomicLongWithDecimal()

    @SerializedName("pow_coinbase_reward")
    val powCoinbaseReward = block.powCoinbaseReward.formatAtomicLongWithDecimal()

    @SerializedName("pop_coinbase_reward")
    val popCoinbaseReward = block.popCoinbaseReward.formatAtomicLongWithDecimal()

    @SerializedName("block_content_metapackage")
    val blockContentMetapackage = BlockContentMetapackageInfo(block.blockContentMetapackage)

    @SerializedName("size")
    val size = block.size

    @SerializedName("header")
    val header = if (number % 20 == 0) {
        BlockUtility.assembleBlockHeader(
            number,
            version,
            previousHash,
            previousKeystoneHash,
            secondPreviousKeystoneHash,
            merkleRoot,
            timestamp,
            encoded_difficulty,
            winningNonce
        ).toHex()
    } else {
        BlockUtility.assembleBlockHeader(
            number,
            version,
            previousHash,
            secondPreviousHash,
            thirdPreviousHash,
            merkleRoot,
            timestamp,
            encoded_difficulty,
            winningNonce
        ).toHex()
    }
}
